AppController/djinn.rb

```diff
@@ -4996,6 +4907,49 @@ def find_lowest_free_port(starting_port)
   end
 
 
+  # Updates @app_info_map with registered instances. This must be called under
+  # APPS_LOCK.
+  def update_registered_instances(version_key)
+    begin
+      zk_instances = ZKInterface.get_children(
+        "/appscale/instances_by_version/#{version_key}")
+    rescue FailedZooKeeperOperationException
+      Djinn.log_warn('Unable to update registered instances.')
+      return
+    end
+
+    unless zk_instances.empty?
+      @app_info_map[version_key] = {} unless @app_info_map.key?(version_key)
+      unless @app_info_map[version_key].key?('appservers')
+        @app_info_map[version_key]['appservers'] = []
+      end
+    end
+    known_instances = @app_info_map[version_key]['appservers']
+
+    # Replace instance assignments with any new registered instances.
+    zk_instances.each { |instance_key|
+      next if known_instances.include?(instance_key)
+
+      # Find and remove an entry for this AppServer node and app.
+      ip = instance_key.split(':')[0]
+      match = known_instances.index("#{ip}:-1")
+      if match
+        known_instances.delete_at(match)
+        known_instances << instance_key
+      else
+        Djinn.log_warn("Ignoring unassigned instance: #{instance_key}.")
+      end
+    }
+
+    # Account for instances that have been stopped.
+    known_instances.delete_if { |instance_key|
+      pending = instance_key.split(':')[-1] == '-1'
+      registered = zk_instances.include?(instance_key)
+      !pending && !registered
```

Comment:
This block is doing what you described as the "other side of the coin" check. If there's a non-pending entry in `@app_info_map` and the entry is not in the list of registered instances that we fetched from ZooKeeper, then we remove the entry from `@app_info_map` because the AppManager has stopped that previously-registered instance.

For example,
1. The shadow assigns an instance to machine1 (machine1:-1)
2. The AppManager on machine one starts the instance and adds a registration node.
3. The AppManager sees that HAProxy believes the instance is down. It stops the instance and removes the registration node.
4. The shadow (in this block) sees that it has an entry in `@app_info_map` that does not exist in the list of registered instances, so it removes the entry from `@app_info_map`.